src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -2636,13 +2715,88 @@ private void setEHinfo(uint EHnumber, ref CORINFO_EH_CLAUSE clause)
+        {
+            _stashedPrecodeFixups.Push(_precodeFixups);
+            _precodeFixups = null;
+            _stashedInlinedMethods.Push(_inlinedMethods);
+            _inlinedMethods = null;
+        }
+
         private void reportInliningDecision(CORINFO_METHOD_STRUCT_* inlinerHnd, CORINFO_METHOD_STRUCT_* inlineeHnd, CorInfoInline inlineResult, byte* reason)
         {
             if (inlineResult == CorInfoInline.INLINE_PASS)
             {
                 // We deliberately ignore inlinerHnd because we have no interest to track intermediate links now.
                 MethodDesc inlinee = HandleToObject(inlineeHnd);
-                _inlinedMethods.Add(inlinee);
+
+                var typicalMethod = inlinee.GetTypicalMethodDefinition();
+                if (!_compilation.CompilationModuleGroup.VersionsWithMethodBody(typicalMethod) &&
+                    typicalMethod is EcmaMethod ecmaMethod &&
+                    ecmaMethod.Module == typicalMethod.Context.SystemModule) // NonVersionable code in System.Collections.Immutable is causing problems and needs to be treated specially
+                {
+                    Debug.Assert(_compilation.CompilationModuleGroup.CrossModuleInlineable(typicalMethod));
+                    if (!typicalMethod.IsNonVersionable())
+                    {
+                        ISymbolNode ilBodyNode = _compilation.SymbolNodeFactory.CheckILBodyFixupSignature(ecmaMethod);
+                        AddPrecodeFixup(ilBodyNode);
+                    }
+
+                    // For cross module inlines, we will compile in a potentially 2 step process
+                    // 1. Compile the method using the set of il bodies available at the start of a multi-threaded compilation run
+                    // 2. If at any time, the set of methods that are inlined includes a method which has an IL body without
+                    //    tokens that are useable in compilation, record that information, and once the multi-threaded portion
+                    //    of the build finishes, it will then compute the IL bodies for those methods, then run the compilation again.
+                    MethodIL methodIL = _compilation.GetMethodIL(typicalMethod);
+                    if (!(methodIL is IMethodTokensAreUseableInCompilation) && methodIL is EcmaMethodIL)
+                    {
+                        _ilBodiesNeeded = _ilBodiesNeeded ?? new List<EcmaMethod>();
+                        _ilBodiesNeeded.Add(ecmaMethod);
+                    }
+                }
+
+                // If during inlining we found Precode fixups, then only if the inline was successful, add them to the set of
+                // fixups that will be used for the entire method
+                List<ISymbolNode> previouslyStashedFixups = _stashedPrecodeFixups.Pop();
+
+                if (_precodeFixups != null)
+                {
+                    previouslyStashedFixups = previouslyStashedFixups ?? new List<ISymbolNode>();
```

Comment:
Non-blocking nit - when <code>previouslyStashedFixups</code> is null, you could directly assign <code>_precodeFixups</code> to it and avoid copying. Not sure how prevalent this is in practice.